### PR TITLE
Add ResultIterator wrapper, `Display` impl for Text, and `get_raw` for Tesseract.

### PR DESCRIPTION
**Pull request by StripedMonkey**
In my own experimentation I ended up running into some issues with null pointers being returned from `GetUTF8Text()` for Tesseract result iterators, and as a result I added a null check to the `AsRef` impl. 

I'm not entirely sure that there should be an AsRef implementation for it in the first place, however, as creating a `CStr` (currently) requires walking the entire string every time. While this isn't the most expensive thing in the world, it's not 0-cost and (probably) a bad idea to conflate, especially since it may be done repeatedly when using them as references instead of once.

In my opinion it should *probably* be changed to an `Into<CStr>` or similar.

### CODE REVIEW

File: src/text.rs [hunk 1]
```diff
@@ -28,6 +30,25 @@ impl Text {
 
 impl AsRef<CStr> for Text {
     fn as_ref(&self) -> &CStr {
```
Comment (ccouzens) [inline src/text.rs:32]: > I'm not entirely sure that there should be an AsRef implementation for it in the first place, however, as creating a CStr (currently) requires walking the entire string every time. While this isn't the most expensive thing in the world, it's not 0-cost and (probably) a bad idea to conflate, especially since it may be done repeatedly when using them as references instead of once.

> In my opinion it should probably be changed to an Into<CStr> or similar.

You're right

https://doc.rust-lang.org/stable/src/core/ffi/c_str.rs.html#260

I'll probably get that changed, but it will be a breaking change.

File: src/text.rs [hunk 2]
```diff
@@ -28,6 +30,25 @@ impl Text {
...
+        if self.0.is_null() {
+            // TODO: This is a dumb hack.
+            // Tesseract may choose to return a null pointer for no text for some invalid states.
```
Comment (ccouzens) [inline src/text.rs:35]: I suppose you've done it this way to avoid a breaking change.

But if we were to release a breaking change, do you see this case being:
1. `Option::None`
2. `Result::Err`
3. Or keep it as an empty string?

**Comment by ccouzens** [comment]
Hey, I'm sorry to have not responded to your PR for several months. I had a horrible life event, and then I lost interest in GitHub for a bit.

File: src/text.rs (hunk 2 above)
Comment (StripedMonkey) [reply src/text.rs:35]: I think a fallible constructor or simply noting the safety requirements on the constructor may be fine, but the current upstream crates make the assumption that it's safe to deref as a cstr so I considered this fixing a safety issue in the current version.

The instances I encountered `Text`s being null were as a result of errors I encountered as setup/config of the image iteration so I'd honestly be more inclined to believe it's an `Err` than anything for the relevant get_text() functions purely on the basis of naming but I'm not sure if there's any other error states that can be encoded other than `null`, so I can see an argument for either.

**Comment by StripedMonkey** [comment]
No worries at all! I'm not paying you :) I hope things get (or are) better for you.

Are you of the opinion that I should change this to a breaking change wrt the `Text` stuff, or would you prefer that be done another time?

File: src/text.rs (hunk 2 above)
Comment (ccouzens) [reply src/text.rs:35]: What about doing something like this to enforce the pointer isn't `null` when `Text` is initialized?

a0926633130fb4e0f7aef1a3a8e1a5f7834f8f4c

It's technically a breaking change.

Comment (StripedMonkey) [reply src/text.rs:35]: Aside from the (IMO) far too wordy `TessBaseApiGetWordStrBoxTextError` naming thing,

 `impl Into<TessBaseApiGetWordStrBoxTextError> for TextNewError` for the associated error types  allows for reducing all the `Text` constructors to
```rust
pub fn get_utf8_text(&mut self) -> Result<Text, TessBaseApiGetUtf8TextError> {
   let ptr = unsafe { TessBaseAPIGetUTF8Text(self.0) }; // Or similar calls
   Text::new(ptr)?
}
```

Comment (ccouzens) [reply src/text.rs:35]: > Aside from the (IMO) far too wordy `TessBaseApiGetWordStrBoxTextError` naming thing (I don't know what `StrBox` is referring to in this situ)

That's basically the error from the [TessBaseAPIGetWordStrBoxText](https://tesseract-ocr.github.io/tessapi/5.x/a00008.html#ab9938845c9b52434ee32a2225aad81cf) method.

There is more documentation about that [here](https://tesseract-ocr.github.io/tessapi/5.x/a02438.html#aa45bc5ee67de0bf66112165ca93399ff). I don't know what a `StrBox` is either, but it probably makes sense to the original c++ developers.

Comment (ccouzens) [reply src/text.rs:35]: `Box` might be referring to a rectangle within the image, rather than heap memory like in Rust.

Comment (ccouzens) [reply src/text.rs:35]: > `impl Into<TessBaseApiGetWordStrBoxTextError> for TextNewError` for the associated error types allows for reducing all the `Text` constructors to
> 
> ```rust
> pub fn get_utf8_text(&mut self) -> Result<Text, TessBaseApiGetUtf8TextError> {
>    let ptr = unsafe { TessBaseAPIGetUTF8Text(self.0) }; // Or similar calls
>    Text::new(ptr)?
> }
> ```

```rust
    pub fn get_utf8_text(&mut self) -> Result<Text, TextNewError> {
        unsafe { Text::new(TessBaseAPIGetUTF8Text(self.0)) }
    }
```

What about this, and just give up on the custom errors? In hindsight there isn't much point having specific errors for all the text functions as they all have the one shared error case (null pointer).

Comment (StripedMonkey) [reply src/text.rs:35]: I'd be for eliminating function-specific errors in this situation, I'm not sure if we can actually provide more "useful" error messages with the API calls at this level. 
Ideally, I'd like to be able to document *why* errors could occur as there's definitely a level of "order of operations" to the API I feel is undocumented/unclear within the C API but I'm also not very fluent in C/C++.

Comment (ccouzens) [reply src/text.rs:35]: I've made the change to stick to a single error for the text conversions here: 21ae92032e489222310722ff8ca60e18f5d7ccc1

I have a branch that includes both our changes https://github.com/ccouzens/tesseract-plumbing/tree/StripedMonkey-main



> I'm not sure if we can actually provide more "useful" error messages with the API calls at this level.

Yeah, I think at this point all we can say is we received a `null` pointer.

> Ideally, I'd like to be able to document _why_ errors could occur as there's definitely a level of "order of operations" to the API I feel is undocumented/unclear within the C API but I'm also not very fluent in C/C++.

Yeah, for example that you actually set an image. The higher level crates that build off of here (`leptess` and `tessearct-rs`), both do that to a degree.

File: src/text.rs (hunk 1 above)
Comment (ccouzens) [reply src/text.rs:32]: Changed in this commit. 93b08ccdd5c782fd058ed05d8db5ede4721d7a50

I didn't go for `Into`, as that is for owned->owned

https://rust-lang.github.io/api-guidelines/naming.html
